File: cpp/graph.cpp
#include "graph.hpp"
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <iostream>

namespace msu_tasks_cpp

{

    using std::string;
    using std::vector;

    Graph::CityId Graph::add_city(const string &title)
    {
        const auto it = titles_to_ids_.find(title);
        if (it != titles_to_ids_.end())
        {
            return it->second;
        }
        else
        {
            const auto city_id = get_new_city_id();
            cities_.emplace(city_id, City(city_id, title));
            titles_to_ids_.emplace(title, city_id);
            return city_id;
        }
    }

    Graph::TransportId Graph::add_transport(const std::string &title)
    {
        TransportId transport_id = -1;
        for (const auto &[key, value] : transports_)
        {
            if (value.title == title)
            {
                transport_id = key;
                break;
            }
        }

        if (transport_id == -1)
        {
            transport_id = get_new_transport_id();
            transports_.emplace(transport_id, Transport(transport_id, title));
        }
        return transport_id;
    }

    void Graph::add_cruise(Cruise &cruise)
    {
        const CruiseId new_cruise_id = get_new_cruise_id();
        cruise.id = new_cruise_id;
        cruises_.emplace(new_cruise_id, cruise);
        get_city(cruise.from_city_id).add_cruise_id(new_cruise_id);
    }

    const std::string Graph::Cruise::to_string(const Graph &graph) const
    {

        const auto city_from = graph.get_cities().at(from_city_id);
        const auto city_to = graph.get_cities().at(to_city_id);

        return city_from.title + " ------> " + city_to.title + " (Время: " + std::to_string(time) + " у. е. | Цена: " + std::to_string(fare) + " ₽)";
    };
}

File: python/graph.py
class City:
    def __init__(self, id, title) -> None:
        self.id = id
        self.title = title

    def add_cruise_id(self, id) -> None:
        new_ids = self.__cruise_ids.copy()
        new_ids.append(id)
        self.__cruise_ids = new_ids

    def __str__(self) -> str: return self.title

    @property
    def cruise_ids(self) -> list[int]: return self.__cruise_ids

    __cruise_ids: list[int] = []


class Transport:
    def __init__(self, id, title) -> None:
        self.id = id
        self.title = title

    def __str__(self) -> str: return self.title


class Cruise:
    def __init__(self, from_city_id, to_city_id, transport_type_id, time, fare, id=0) -> None:
        self.id = id
        self.from_city_id = from_city_id
        self.to_city_id = to_city_id
        self.transport_type_id = transport_type_id
        self.time = time
        self.fare = fare

    def __str__(self) -> str: return '(' + str(self.id) + ', ' + \
        str(self.from_city_id) + ', ' + str(self.to_city_id) + ')'


class Graph:

    def add_city(self, title) -> int:
        city_id = -1

        for key, value in self.__cities.items():
            if value.title == title:
                city_id = key
                break

        if city_id == -1:
            city_id = self.__get_new_city_id()
            self.__cities[city_id] = City(city_id, title)

        return city_id

    def add_cruise(self, cruise: Cruise) -> None:
        new_cruise_id = self.__get_new_cruise_id()
        cruise.id = new_cruise_id
        self.__cruises[new_cruise_id] = cruise
        self.__cities[cruise.from_city_id].add_cruise_id(new_cruise_id)

    def add_transport(self, title) -> int:
        transport_id = -1
        for key, value in self.__transports.items():
            if value.title == title:
                transport_id = key
                break

        if transport_id == -1:
            transport_id = self.__get_new_transport_id()
            self.__transports[transport_id] = Transport(transport_id, title)

        return transport_id

    def get_city_cruises(
        self, city_id) -> list[int]: return self.__cities[city_id].cruise_ids()

    def get_city(self, id) -> City: return self.__cities[id]

    def get_cruise(self, id) -> Cruise: return self.__cruises[id]

    def get_transport(self, id) -> Transport: return self.__transports[id]

    @property
    def cities(self) -> dict[int, City]: return self.__cities

    @property
    def cruises(self) -> dict[int, Cruise]: return self.__cruises

    @property
    def transports(self) -> dict[int, Transport]: return self.__transports

    def get_cities_amount(self) -> int: return self.__cities.size()

    def get_cruises_amount(self) -> int: return self.__cruises.size()

    __cities: dict[int, City] = {}
    __cruises: dict[int, Cruise] = {}
    __transports: dict[int, Transport] = {}
    __city_id_counter: int = 0
    __cruise_id_counter: int = 0
    __transport_id_counter: int = 0

    def __get_new_city_id(self) -> int:
        result = self.__city_id_counter
        self.__city_id_counter += 1
        return result

    def __get_new_cruise_id(self) -> int:
        result = self.__cruise_id_counter
        self.__cruise_id_counter += 1
        return result

    def __get_new_transport_id(self) -> int:
        result = self.__transport_id_counter
        self.__transport_id_counter += 1
        return result
